File: nodejs/checkio/src/home/sort_array_by_element_frequency.ts
import assert from "assert";

function frequencySort<T extends number | string>(items: T[]): T[] {
	const frequencyMap: Map<T, number> = new Map();

	for (const item of items) {
		frequencyMap.set(item, (frequencyMap.get(item) || 0) + 1);
	}

	const sorted = Array.from(frequencyMap.entries()).sort((a, b) => {
		return b[1] - a[1];
	});

	return sorted.flatMap(([item, freq]) => Array(freq).fill(item));
}

console.log("Example:");
console.log(JSON.stringify(frequencySort([4, 6, 2, 2, 6, 4, 4, 4])));

// These "asserts" are used for self-checking
assert.deepStrictEqual(
	frequencySort([4, 6, 2, 2, 6, 4, 4, 4]),
	[4, 4, 4, 4, 6, 6, 2, 2],
);
assert.deepStrictEqual(
	frequencySort([4, 6, 2, 2, 2, 6, 4, 4, 4]),
	[4, 4, 4, 4, 2, 2, 2, 6, 6],
);
assert.deepStrictEqual(frequencySort(["bob", "bob", "carl", "alex", "bob"]), [
	"bob",
	"bob",
	"bob",
	"carl",
	"alex",
]);
assert.deepStrictEqual(frequencySort([17, 99, 42]), [17, 99, 42]);
assert.deepStrictEqual(frequencySort([]), []);
assert.deepStrictEqual(frequencySort([1]), [1]);

console.log("Coding complete? Click 'Check Solution' to earn rewards!");

File: python/checkio/home/sort-array-by-element-frequency.py
from collections import Counter
from collections.abc import Iterable


def frequency_sort(items: list[str | int]) -> Iterable[str | int]:
    freq = Counter(items)
    return sorted(items, key=lambda x: (-freq[x], items.index(x)))


print("Example:")
print(list(frequency_sort([4, 6, 2, 2, 6, 4, 4, 4])))

# These "asserts" are used for self-checking
assert list(frequency_sort([4, 6, 2, 2, 6, 4, 4, 4])) == [4, 4, 4, 4, 6, 6, 2, 2]
assert list(frequency_sort([4, 6, 2, 2, 2, 6, 4, 4, 4])) == [4, 4, 4, 4, 2, 2, 2, 6, 6]
assert list(frequency_sort(["bob", "bob", "carl", "alex", "bob"])) == [
    "bob",
    "bob",
    "bob",
    "carl",
    "alex",
]
assert list(frequency_sort([17, 99, 42])) == [17, 99, 42]
assert list(frequency_sort([])) == []
assert list(frequency_sort([1])) == [1]

print("The mission is done! Click 'Check Solution' to earn rewards!")
